### PR TITLE
Fixing SCTE-35 implicit signal closing rules.

### PR DESCRIPTION
**Pull request by mikereedell**
- Treat the open events slice like a stack, with the oldest event on the bottom since the
logic that closes events iterates from the tail of the slice to the head.
- Added unit test that illustrated this issue and verifies the fix.

### CODE REVIEW

**Comment by markniebur** [comment]
I haven't forgotten about this... will get to it tomorrow.

File: scte35/segmentationdescriptor.go [hunk 1]
```diff
@@ -255,19 +257,20 @@ func (d *segmentationDescriptor) CanClose(out SegmentationDescriptor) bool {
...
 		}
 	case segCloseEventIDNotNested:
+		// this should also consider segnum == segexpected for IN signals closing an out signal.
+		if d.IsIn() && d.EventID() == out.EventID() && d.SegmentNumber() == d.SegmentsExpected() {
+			return true
+		}
 		if d.EventID() != out.EventID() {
 			return false
 		}
 		fallthrough
```
Comment (markniebur) [inline scte35/segmentationdescriptor.go]: Is this fallthrough still needed?  From cog esam 2.0 doc, it seems like the logic has diverged enough that only the first if block is needed and all the old code can be deleted.

File: scte35/segmentationdescriptor.go [hunk 2]
```diff
@@ -255,19 +257,20 @@ func (d *segmentationDescriptor) CanClose(out SegmentationDescriptor) bool {
...
-	case segCloseNotNested:
-		if d.IsIn() {
-			// if in, last desc is x/x
-			if d.segNum == d.segsExpected {
+	case segCloseNotNested: // only applies to 0x34 and 0x36 with subsegments.
+		if d.IsOut() && (d.TypeID() == SegDescProviderPOStart || d.TypeID() == SegDescDistributorPOStart) {
```
Comment (markniebur) [inline scte35/segmentationdescriptor.go]: I feel like this should be dictated by the closeRules map only 0x34 and 0x36 should have segCloseNotNested, the extra logic here is unneeded, and two places will have to be updated if another signal gets this close rule footnote.

File: scte35/segmentationdescriptor.go [hunk 3]
```diff
@@ -255,19 +257,20 @@ func (d *segmentationDescriptor) CanClose(out SegmentationDescriptor) bool {
...
+			if d.HasSubSegments() && d.SubSegmentNumber() == d.SubSegmentsExpected() {
 				return true
 			}
-		} else if d.IsOut() {
-			// if out, first descriptor in set closes existing open
-			if d.segNum == 1 {
+			if d.TypeID() == out.TypeID() {
```
Comment (markniebur) [inline scte35/segmentationdescriptor.go]: Same story here, I think this is already done in the close rules map and this is just extra, unneeded logic.

File: scte35/segmentationdescriptor.go (hunk 1 above)
Comment (mikereedell) [reply scte35/segmentationdescriptor.go]: The `if d.EventId()...` statement is extraneous.  As is the `fallthrough`.

File: scte35/segmentationdescriptor.go (hunk 2 above)
Comment (mikereedell) [reply scte35/segmentationdescriptor.go]: Good call.

File: scte35/segmentationdescriptor.go (hunk 3 above)
Comment (mikereedell) [reply scte35/segmentationdescriptor.go]: Agreed.

**Pull request merged**